pkg/controller/cloud/pvlcontroller.go

```diff
@@ -205,9 +205,28 @@ func (pvlc *PersistentVolumeLabelController) createPatch(vol *v1.PersistentVolum
 	if newVolume.Labels == nil {
 		newVolume.Labels = make(map[string]string)
 	}
+	if len(volLabels) != 0 {
+		if newVolume.Spec.NodeAffinity == nil {
+			newVolume.Spec.NodeAffinity = new(v1.VolumeNodeAffinity)
+		}
+		if newVolume.Spec.NodeAffinity.Required == nil {
+			newVolume.Spec.NodeAffinity.Required = new(v1.NodeSelector)
+		}
+	}
+	requirements := make([]v1.NodeSelectorRequirement, len(volLabels))
+	i := 0
 	for k, v := range volLabels {
 		newVolume.Labels[k] = v
+		// Set NodeSelectorRequirements based on the labels
+		requirements[i] = v1.NodeSelectorRequirement{Key: k, Operator: v1.NodeSelectorOpIn, Values: []string{v}}
+		i = i + 1
 	}
+	// Populate NodeAffinity with requirements
+	if len(volLabels) != 0 {
+		t := v1.NodeSelectorTerm{MatchExpressions: requirements}
+		newVolume.Spec.NodeAffinity.Required.NodeSelectorTerms = append(newVolume.Spec.NodeAffinity.Required.NodeSelectorTerms, t)
+	}
+
```

Comment:
We should make sure these labels passed in are the ones given by the cloud provider.

You also bring up a good point. If the user already specified a NodeAffinity in the PV, we should probably not override it.

Reply:
Good point indeed. I did think about the above but thought overriding is(/was) preferred based on this unit-test: https://github.com/kubernetes/kubernetes/blob/master/plugin/pkg/admission/storage/persistentvolume/label/admission_test.go#L161

Reply:
I discussed this briefly with @saad-ali and @verult.  We think it's better if there's any conflict between user provided and cloud provider provided values, then we should error instead of blindly overwriting.  Users should know that they made a configuration mistake because it could indicate that they are accidentally using the wrong volume than intended.

Reply:
Sounds good. Thinking about how to detect conflicts and "merge" with an arbitrary pre-specified affinity in a safe way, the following scheme should work. Thoughts?

```
Scan through each NodeSelectorTerm in NodeAffinity.Required:
  Create a map <NodeSelectorRequirement, bool> populated with NodeSelectorRequirement s corresponding to cloud specific labels to record if a matching NodeSelectorRequirement has been found in current NodeSelectorTerm
    Scan through each NodeSelectorRequirement in the NodeSelectorTerm:
        If NodeSelectorRequirement.Key matches but operator or values do not match, return error due to conflict.
        If key, operator, value of NodeSelectorRequirement matches entry in map, mark it in map.
    Append NodeSelectorRequirement entries from map that have not yet been marked/matched with NodeSelectorRequirements (corresponding to cloud specific label) in current NodeSelectorTerm
``` 

This should result in the NodeSelectorRequirement entries corresponding to cloud specific labels being "and"ed to each NodeSelectorTerm entry of an arbitrarily specified NodeAffinity.Required block.

Reply:
This could work; alternatively maybe we just want to leave the NodeAffinity alone if the user already specified something?